src/deluge/gui/ui/keyboard/layout/isomorphic.cpp

```diff
@@ -20,27 +20,100 @@
 #include "gui/ui/audio_recorder.h"
 #include "gui/ui/browser/sample_browser.h"
 #include "gui/ui/sound_editor.h"
+#include "model/instrument/melodic_instrument.h"
+#include "model/model_stack.h"
 #include "model/settings/runtime_feature_settings.h"
 #include "util/functions.h"
 #include <limits>
 
+#define VEL_COL kDisplayWidth
+#define MOD_COL (kDisplayWidth + 1)
+
 namespace deluge::gui::ui::keyboard::layout {
 
+static inline void popupUint8(deluge::hid::Display* display, uint8_t val) {
+	char valStr[4] = {0};
+	intToString(val, valStr, 1);
+	display->displayPopup(valStr);
+}
+
 void KeyboardLayoutIsomorphic::evaluatePads(PressedPad presses[kMaxNumKeyboardPadPresses]) {
 	uint8_t noteIdx = 0;
 
 	currentNotesState = NotesState{}; // Erase active notes
 
+	velocityMinHeld = false;
+	velocityMaxHeld = false;
+
+	modMinHeld = false;
+	modMaxHeld = false;
+
 	for (int32_t idxPress = 0; idxPress < kMaxNumKeyboardPadPresses; ++idxPress) {
-		if (presses[idxPress].active && presses[idxPress].x < kDisplayWidth) {
-			currentNotesState.enableNote(noteFromCoords(presses[idxPress].x, presses[idxPress].y),
-			                             getDefaultVelocity());
+		auto pressed = presses[idxPress];
+		if (pressed.active) {
+			// in note columns
+			if (pressed.x < kDisplayWidth) {
+				currentNotesState.enableNote(noteFromCoords(pressed.x, pressed.y), velocity);
+			}
+			else if (pressed.x == VEL_COL) { // in velocity columns (audition pads)
+				velocity32 = velocityMin + pressed.y * velocityStep;
+				velocity = velocity32 >> kVelModShift;
+				popupUint8(display, velocity);
+				if (pressed.y == 0) {
+					velocityMinHeld = true;
+				}
+				else {
+					velocityMaxHeld = true;
+				}
+			}
+			else if (pressed.x == MOD_COL) { // in mod columns (audition pads)
+				mod32 = modMin + pressed.y * modStep;
+				char modelStackMemory[MODEL_STACK_MAX_SIZE];
+				ModelStack* modelStack = setupModelStackWithSong(modelStackMemory, currentSong);
+				ModelStackWithTimelineCounter* modelStackWithTimelineCounter =
+				    modelStack->addTimelineCounter(currentSong->currentClip);
```

Comment:
Is there already model stack somewhere in the class hierarchy above here? I couldn't find one.

Reply:
I don't think so, this is probably the best way to get it